### PR TITLE
Add test coverage for shortint layer

### PR DESCRIPTION
**Pull request by soonum**
This PR, once completed, will lead to test coverage implementation for the other layers.

To do before merging:

- [x] Add specific target for generating keys to populate keycache
- [x] Create and test workflow dedicated to coverage

### CODE REVIEW

File: tfhe/src/keycache/mod.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub mod utils {
     }
 
     pub trait NamedParam {
-        fn name(&self) -> &'static str;
+        fn name(&self) -> String;
```
Comment (IceTDrinker) [inline tfhe/src/keycache/mod.rs:24]: why ?

File: tfhe/src/shortint/keycache.rs [hunk 2]
```diff
@@ -396,7 +438,78 @@ impl KeycacheWopbsV0 {
...
+            })
+        })
+    }
+}
+
+impl NamedParam for KeySwitchingKeyParams {
+    fn name(&self) -> String {
+        format!("{}__{}__{}", self.0.name(), self.1.name(), self.2.name())
+    }
+}
```
Comment (IceTDrinker) [inline tfhe/src/shortint/keycache.rs:472]: can this macroized ?

Comment (IceTDrinker) [reply tfhe/src/shortint/keycache.rs:472]: using paste and the exposed constants for parameter structs ?

File: tfhe/src/keycache/mod.rs (hunk 1 above)
Comment (tmontaigu) [reply tfhe/src/keycache/mod.rs:24]: Some tests (KeySwitchingKey test if I remember correctly )generates parameters at runtime, so parameters for which there is no associated constant, so no associated &static str.

So the NamedParam had to be changed to allow returning some string built at runtime

Comment (IceTDrinker) [reply tfhe/src/keycache/mod.rs:24]: huh, ok, that's too bad, do you remember where this was ?

Comment (tmontaigu) [reply tfhe/src/keycache/mod.rs:24]: I think this one is an example https://github.com/zama-ai/tfhe-rs/blob/95dcf95e88299910a51f9f8805786327f5e10127/tfhe/src/shortint/key_switching_key/test.rs#L55

Comment (IceTDrinker) [reply tfhe/src/keycache/mod.rs:24]: arf, indeed

we are hitting the problem as the keys now go through the cache ? or is it because we bench them now ?

File: tfhe/src/keycache/mod.rs [hunk 3]
```diff
@@ -38,8 +38,9 @@ pub mod utils {
             named_params_impl!(expose $($const_param),*);
 
             impl NamedParam for $param_type {
-                fn name(&self) -> &'static str {
+                fn name(&self) -> String {
                     named_params_impl!({*self; $param_type} == ( $($const_param),* ));
+                    panic!("Unnamed parameters");
```
Comment (IceTDrinker) [inline tfhe/src/keycache/mod.rs:43]: why this change ? having the panic in the macro which expands to the implementation seemed better ?

Comment (IceTDrinker) [reply tfhe/src/keycache/mod.rs:43]: nevermind, just saw this in another part of the code

we need to have that chat about keycache

File: tfhe/src/keycache/mod.rs (hunk 1 above)
Comment (tmontaigu) [reply tfhe/src/keycache/mod.rs:24]: Yes, to speed up coverage all keys are fetched from the keycache, and some of these key have parameters which are made at runtime and do not have as `const ...`

Comment (IceTDrinker) [reply tfhe/src/keycache/mod.rs:24]: is it expected ? or should we have static params down the line ? i.e. is it a workaround at the keyswitching level or is it the normal flow that is expected later ?

Comment (tmontaigu) [reply tfhe/src/keycache/mod.rs:24]: I think we did that because for some of the parameters of keyswitching, it did not make sense to add the `const ...`

**Review by IceTDrinker — COMMENTED** [review]
Comments, some of them may not be relevant but I'm not sure if some of those decisions are concerted or not

File: tfhe/src/shortint/keycache.rs [hunk 4]
```diff
@@ -240,29 +241,46 @@ named_params_impl!( ShortintParameterSet =>
...
+            {
+                *self;
+                ShortintKeySwitchingParameters
+            } == (PARAM_KEYSWITCH_1_1_KS_PBS_TO_2_2_KS_PBS)
+        );
+
+        format!(
+            "PARAM_KEYSWITCH_CUSTOM_KS_LEVEL_{}_KS_BASE_LOG_{}",
+            self.ks_level.0, self.ks_base_log.0
+        )
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/src/shortint/keycache.rs:280]: Yes, for the custom KSK parameters built on the fly.

File: tfhe/examples/utilities/generates_test_keys.rs [hunk 5]
```diff
@@ -17,56 +27,71 @@ fn client_server_keys() {
...
-
-            println!("Generation took {stop} seconds");
+        generate_pbs_multi_bit_keys(&ALL_MULTI_BIT_PARAMETER_VEC);
+    } else if coverage_only {
+        println!("Generating shortint (ClientKey, ServerKey) for coverage");
+
+        const PBS_PARAMS: [ClassicPBSParameters; 5] = [
+            PARAM_MESSAGE_1_CARRY_1_KS_PBS,
+            PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+            PARAM_MESSAGE_3_CARRY_3_KS_PBS,
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/examples/utilities/generates_test_keys.rs:55]: Yes we use it for KSK tests.

File: tfhe/examples/utilities/generates_test_keys.rs [hunk 6]
```diff
@@ -17,56 +27,71 @@ fn client_server_keys() {
...
+            ShortintKeySwitchingParameters,
+        ); 2] = [
+            (
+                PARAM_MESSAGE_1_CARRY_1_KS_PBS,
+                PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+                PARAM_KEYSWITCH_1_1_KS_PBS_TO_2_2_KS_PBS,
+            ),
+            (
+                PARAM_MESSAGE_1_CARRY_1_KS_PBS,
+                PARAM_MESSAGE_3_CARRY_3_KS_PBS,
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/examples/utilities/generates_test_keys.rs:81]: It is used in the test named `gen_multi_keys_test_add_with_overflow()`

File: Makefile [hunk 7]
```diff
@@ -80,6 +87,12 @@ install_node:
 	$(SHELL) -i -c 'nvm install node' || \
 	( echo "Unable to install node, unknown error." && exit 1 )
 
+.PHONY: install_tarpaulin # Install tarpaulin to perform code coverage
+install_tarpaulin: install_rs_build_toolchain
+	@cargo tarpaulin --version > /dev/null 2>&1 || \
+	cargo $(CARGO_RS_BUILD_TOOLCHAIN) install cargo-tarpaulin || \
```
[the remark this answers is not in the document]
Comment (soonum) [reply Makefile:93]: Why would you lock the tarpaulin install ?

File: tfhe/examples/utilities/generates_test_keys.rs [hunk 8]
```diff
@@ -87,27 +112,112 @@ fn client_server_keys() {
...
+        ClassicPBSParameters,
+        ClassicPBSParameters,
+        ShortintKeySwitchingParameters,
+    )],
+) {
+    println!("Generating shortint Key Switching keys (ClientKey, ServerKey)");
+
+    for (i, (param_1, param_2, param_ksk)) in params.iter().copied().enumerate() {
+        println!(
+            "Generating [{} / {}] : {}__{}__{}",
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/examples/utilities/generates_test_keys.rs:178]: If we use KSK params name we could have collisions. For example we have a test that uses 1_1 and 2_2 with 1_1_ksk_to_2_2, and then we have a second test that uses 1_1 and 3_3 with the same KSK parameters.
Plus for the KSK parameters built on the fly, we cannot pass an existing const parameters set to the keycache method.

File: tfhe/src/shortint/server_key/tests/shortint.rs [hunk 9]
```diff
@@ -2298,21 +2360,13 @@ where
...
-        for _ in 0..30 {
-            ct_res = sks.mul_lsb(&ct_res, &ctxt_0);
-            clear = (clear * clear_0) % modulus;
-
-            // decryption of ct_res
-            let dec_res = cks.decrypt(&ct_res);
+        let dec_res = cks.decrypt(&ct_res);
 
-            // assert
-            assert_eq!(clear, dec_res);
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: We don't need this step in default operation, since there is a full-propagate at the beginning and the end of the operation.
This must be a copy-paste error while creating the test.

File: tfhe/src/shortint/server_key/tests/shortint.rs [hunk 10]
```diff
@@ -537,7 +601,7 @@ where
 
     let modulus = cks.parameters.message_modulus().0 as u64;
 
-    for _ in 0..10 {
+    for _ in 0..NB_TEST {
```
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/src/shortint/server_key/tests/shortint.rs:604]: Some tests would use 10 and other 30. Naively I went for the bigger number knowing that could increase test duration.
But maybe there is a rationale behind choosing 10 over 30 in some places ?

File: tfhe/src/shortint/keycache.rs (hunk 2 above)
Comment (tmontaigu) [reply tfhe/src/shortint/keycache.rs:472]: This can't be macroized as some parameters (mainly ShortintKeySwitchingParameters, (self.2.name()) may be computed at runtime

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: default tests have  two things to test in order to go into all branches possible

- test the operation when inputs are clear of carries
- test the operation when inputs do have carries

The second inner loop that was removed tested the first case (input with clean carries
which the new test still test.

 The comment `//add multiple times to raise the degree` was wrong, their was no degree being raised because it used the default ops

[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: The difference is that instead of testing 10*30 ops we are now testing NB_TEST ops

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: wasn't the inner loop for carry stuff ? In any case I was surprised if it was meant for coverage only

File: tfhe/examples/utilities/generates_test_keys.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/examples/utilities/generates_test_keys.rs:55]: ok

File: Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply Makefile:93]: When installing binaries it's better to install with the committed lockfile as it's the tested configuration by their CI I believe

File: tfhe/examples/utilities/generates_test_keys.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/examples/utilities/generates_test_keys.rs:81]: well I see 1_1 and 2_2 in that test not 1_1 and 3_3 ? 🤔

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: The comment says the inner loop is for raising degree, testing case where carries are non empty.

However the code uses `mul_lsb` in that case (in shortint_default_add, the inner loop uses `add`). And since `mul_lsb` and `add` are default ops, then they output result that do not have any carries, so we aren't really testing the non-clean carry case

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs:2314]: Do you want to sync with @soonum on that to be sure everything is in order ?

File: tfhe/examples/utilities/generates_test_keys.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (soonum) [reply tfhe/examples/utilities/generates_test_keys.rs:81]: I think we are not on the same page:
```rust
#[test]
fn gen_multi_keys_test_add_with_overflow() {
    let keys = KEY_CACHE_KSK.get_from_param((
        PARAM_MESSAGE_1_CARRY_1_KS_PBS,
        PARAM_MESSAGE_3_CARRY_3_KS_PBS,
        PARAM_KEYSWITCH_1_1_KS_PBS_TO_2_2_KS_PBS,
    ));
    let (ck1, sk1) = (keys.client_key_1(), keys.server_key_1());
    let (ck2, sk2) = (keys.client_key_2(), keys.server_key_2());
    let ksk = keys.key_switching_key();

```

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/examples/utilities/generates_test_keys.rs:81]: must be in the wrong branch then, my bad

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs]: Provided the test with dirty carries are added back in the next coverage PR as discussed the loop/test behavior change is ok

File: tfhe/Cargo.toml [hunk 11]
```diff
@@ -120,6 +120,9 @@ generator_aarch64_aes = ["concrete-csprng/generator_aarch64_aes"]
 
 # Private features
 __profiling = []
+__coverage = []
+
+coverage = []
```
Comment (IceTDrinker) [inline tfhe/Cargo.toml:125]: please remove the public feature and keep the private one, switch all usage of the public feature to the private one

File: tfhe/src/shortint/server_key/tests/shortint.rs [hunk 12]
```diff
@@ -4,9 +4,29 @@ use paste::paste;
 use rand::Rng;
 
 /// Number of assert in randomized tests
-const NB_TEST: usize = 30;
+#[cfg(not(feature = "coverage"))]
+const NB_TEST: usize = 200;
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/tests/shortint.rs:8]: may be worth a try is

```rust
const NB_TEST: usize = {
    #[cfg(not(feature = "__coverage"))]
    200
    #[cfg(feature = "__coverage")]
    2
};
```

legal rust ? (to keep things locally)

also use the private feature

I don't remember if I already asked why 200 ? that seems like a lot ? Or is it linked to getting the loop out of the default tests and adding separate dirty inputs test ?

File: Makefile [hunk 13]
```diff
@@ -304,6 +317,12 @@ test_shortint: install_rs_build_toolchain
 	RUSTFLAGS="$(RUSTFLAGS)" cargo $(CARGO_RS_BUILD_TOOLCHAIN) test --profile $(CARGO_PROFILE) \
 		--features=$(TARGET_ARCH_FEATURE),shortint,internal-keycache -p tfhe -- shortint::
 
+.PHONY: test_shortint_cov # Run the tests of the shortint module with code coverage
+test_shortint_cov: install_tarpaulin
+	RUSTFLAGS="$(RUSTFLAGS)" cargo $(CARGO_RS_CHECK_TOOLCHAIN) tarpaulin --profile $(CARGO_PROFILE) \
+		--out Xml --output-dir coverage/shortint --line --engine Llvm --timeout 500 \
```
[the remark this answers is not in the document]
Comment (soonum) [reply Makefile:323]: By default tarpaulin create a cobertura.xml in the current working directory. To avoid any overwrite with future layers of coverage, I decided to create a `coverage` directory that groups reports by layers.

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 12 above)
Comment (soonum) [reply tfhe/src/shortint/server_key/tests/shortint.rs:8]: Yeah 200 seems a lot. I'll switch back to 30 and we'll use that values for the to be implemented dirty inputs tests.

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/server_key/tests/shortint.rs]: not "added back" but "added", the parts that were removed did not test dirty carry behavior

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs]: 😵‍💫

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 12 above)
Comment (soonum) [reply tfhe/src/shortint/server_key/tests/shortint.rs:8]: Sadly, you're proposal is not legal Rust.

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs]: ok then I misunderstood the whole conversation at some point I guess

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 12 above)
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs:8]: 😭

File: tfhe/src/shortint/server_key/tests/shortint.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/tests/shortint.rs]: yep makes sense now, all good, thanks !

File: Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply Makefile:323]: I mean is it in the .gitignore file ?

**Review by IceTDrinker — APPROVED** [review]
Alright let's go ! Thanks a lot for the work 🙏

File: Makefile (hunk 13 above)
[the remark this answers is not in the document]
Comment (soonum) [reply Makefile:323]: Nope it's not as for now. Should I add it now since the approval test suite just begun ?

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply Makefile:323]: yes please, updating the gitignore file is fine, we'll wait for https://github.com/zama-ai/tfhe-rs/pull/498/checks?check_run_id=16707790649 to complete (integer tests)

**Comment by soonum** [comment]
@slab-ci cpu_wasm_test

**Review by IceTDrinker — APPROVED** [review]
